common-npm-packages/webdeployment-common/msdeployutility.ts

```diff
@@ -254,111 +250,91 @@ export async function getMSDeployFullPath(): Promise<string> {
-    return new Promise<string>((resolve, reject) => {
-        var regKey = new winreg({
+export async function installedMSDeployVersionSupportsTokenAuth(): Promise<boolean | undefined> {
+    // MSDeploy 9.0.7225 is the first product version to support token auth
+    const minimalMSDeployVersion = "9.0.7225";
+    const msDeployVersion = await getInstalledMSDeployVersion();
+    if (!msDeployVersion) {
+        tl.debug('Could not determine MSDeploy version. Assuming it is not installed.');
+        return undefined;
+    }
+    return semver.gte(semver.coerce(msDeployVersion), semver.coerce(minimalMSDeployVersion));
+}
+
+function getMSDeployLatestRegKey(): Promise<winreg.Registry> {
+    return new Promise<winreg.Registry>((resolve, reject) => {
+
+        const msdeployRegistryPath = "\\SOFTWARE\\Microsoft\\IIS Extensions\\MSDeploy";
+        const regKey = new winreg({
             hive: winreg.HKLM,
-            key: registryKey
-        })
+            key: msdeployRegistryPath
+        });
 
-        regKey.values(function (err, items: { name: string, value: string }[]) {
+        regKey.keys(function (err, subRegKeys) {
             if (err) {
                 reject(tl.loc("UnabletofindthelocationofMSDeployfromregistryonmachineError", err));
+                return;
             }
 
-            if (shouldUseMSDeployTokenAuth()) {
-                const versionItem = items.find(item => item.name === "Version");
-                if (!versionItem) {
-                    reject(tl.loc("MissingMSDeployVersionRegistryKey"));
-                }
+            tl.debug(`Found ${subRegKeys.length} subkeys under ${msdeployRegistryPath}`);
 
-                const minimalSupportedVersion = "9.0.7225.0";
-                const version = versionItem.value;
-                tl.debug(`Installed MSDeploy Version: ${version}`);
+            let latestKeyVersion = 0;
+            let latestSubKey: winreg.Registry;
 
-                // MSDeploy 9.0.7225.0 is the first version to support token auth
-                if (compareVersions(version, minimalSupportedVersion) < 0) {
-                    reject(tl.loc("UnsupportedMSDeployVersion", version));
+            for (var index in subRegKeys) {
+                const subRegKey = subRegKeys[index].key;
+                tl.debug("Found subkey " + subRegKey);
+                const subKeyVersion = subRegKey.substring(subRegKey.lastIndexOf("\\") + 1);
+                const subKeyVersionNumber = parseFloat(subKeyVersion);
+                if (!isNaN(subKeyVersionNumber) && subKeyVersionNumber > latestKeyVersion) {
+                    latestKeyVersion = subKeyVersionNumber;
+                    latestSubKey = subRegKeys[index];
                 }
             }
-
-            const installPathItem = items.find(item => item.name === "InstallPath");
-            if (!installPathItem) {
-                reject(tl.loc("MissingMSDeployInstallPathRegistryKey"));
+            if (latestKeyVersion < 3) {
```

Comment:
Why does the key have to be less than 3? I somehow lost the logic of this piece of code.

Reply:
This logic is iterating through all installed versions of msdeploy in registry. Details about every version are stored under separate sub key (see below). In my case I have only one version. But in case there are multiple versions the highest one which is greater than 3 should be picked. I believe that versions older than 3 are not compatible either with our tasks or with Azure App Service API. 

![image](https://github.com/microsoft/azure-pipelines-tasks-common-packages/assets/5574525/b1365a1c-9038-4f10-ac54-0f778b94ee9f)

Reply:
Okay, please elaborate on this somewhere in comments, otherwise it's impossible to know the logic behind this code in the future.

Reply:
Fixed. Thank you for your review.